[JSC] Wasm test regressions on 32-bit ARM after 256472@main

aoikonomopoulos opened the pull request:
#### 24455f01485116bf93e8b750e86691c44d70230d
<pre>
[JSC] Wasm test regressions on 32-bit ARM after 256472@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=247886">https://bugs.webkit.org/show_bug.cgi?id=247886</a>

Reviewed by Yusuke Suzuki.

Testing if the index is &lt; 0 is not appropriate on 32-bit ARM.
Define and use minTypeValue instead.

Implementing <a href="https://bugs.webkit.org/show_bug.cgi?id=247454">https://bugs.webkit.org/show_bug.cgi?id=247454</a> would be a
better medium-term fix.

* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::Type::dump const):
* Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:

Canonical link: <a href="https://commits.webkit.org/257116@main">https://commits.webkit.org/257116@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7766d0bc915a5473af81f9e1e671f2534b613ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7183 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107433 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167706 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7648 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35957 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90450 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/104061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103613 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84572 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/32806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88838 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1163 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84544 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1145 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28585 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4892 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5979 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87380 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2431 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41699 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19604 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -469,7 +469,7 @@ inline void Type::dump(PrintStream& out) const
     if (index != TypeDefinition::invalidIndex) {
         auto signedIndex = static_cast<std::make_signed<TypeIndex>::type>(index);
-        if (signedIndex < 0) {
+        if ((signedIndex < 0) && (signedIndex >= minTypeValue)) {
```
Why is it different in 32bit? I think this is fundamental invariant, and having difference in 32bit sounds wrong, and the right fix is fixing 32bit not to have this wrong condition.
Can you explain the reason of 32bit difference?

jjgriego replied:
(sorry for barging in but timezone differences may make this more expedient) as I understand, the problem is that on 64-bit unix/linux (and even windows, apparently); user-space pointers are never going to have the sign bit set, so we can directly pun them as `intptr_t` and be assured the result is going to be positive

but on 32-bit linux, there's no such restriction: user-space pointers can, in principle, have the MSB set [1], so we need something different as a test for "is this a TypeDefinition* or a TypeKind"

... though I agree fixing the linked bug would be a more robust solution

[1] https://linux-kernel-labs.github.io/refs/heads/master/lectures/address-space.html#linux-address-space

aoikonomopoulos replied:
On top of the address space split being configurable, it is very much the case that userland pointers may have the MSB set in the 32-bit-userland-running-on-a-64-bit-kernel used by on the ARM buildbots (necessary to take advantage of more powerful ARM hardware). In fact, userspace gets a separate address space: http://sprunge.us/zpMowI. Not sure if they're doing that to mitigate spectre or for some other reason.

Constellation replied:
That makes sense. So I think the right fix is that we should use first 0-4KB pointers for these types (probably not using 0 since it should be used for null / invalid). The first page for address 0-4KB is ensured to be unmapped in C language standard, and all modern OSes ensure this (otherwise, nullptr can point at some memory). So these 0-4KB pointers can be used reliably.
Right now, TypeKind is using negative values. But we should just use small integer values, and then, we can check this on 64bit and 32bit platforms.

```
if (value < 4KB)
    this is TypeKind
```
I think that removing 32bit specific hack is super important, and in this case, I think this is the right fix.

aoikonomopoulos replied:
I understand that type indices being invalid pointers is desirable. However, the current values of type indices are not arbitrary -- they're the [binary representation](https://www.w3.org/TR/wasm-core-1/#types%E2%91%A6) of the corresponding wasm types.

I see two paths forward:

1. Flip the sign of the runtime representation of wasm types when loading them. This feels like introducing more opportunity for errors.
2. Use the C++ type system to avoid errors in the interpretation of the type index (https://bugs.webkit.org/show_bug.cgi?id=247454)

Option 2 looks like the clear winner here, but maybe you have some better idea.

In any case, this feels like a "nice to have" on 32-bits. The type index representation is still invalid as a pointer on 64-bits, it's only 32-bits that doesn't have this property. With that in mind, I'd rather land this simple fix now to keep the tests green and solve this by option 2 above at a later stage, what do you think?

Constellation replied:
I think it is OK to use the current simple check. But right now, this patch is only changing it in `Type::dump`, is it the only place?
Also, if 32bit has this kind of problem, then `(signedIndex < 0) && (signedIndex >= minTypeValue)` check should be extracted as a function.

aoikonomopoulos replied:
OK, did that and updated the PR. AFAICT this is the only site where the code checks for < 0.

aoikonomopoulos replied:
Ping.

Constellation reviewed (approved):
r=me

aoikonomopoulos committed:
[JSC] Wasm test regressions on 32-bit ARM after 256472@main
https://bugs.webkit.org/show_bug.cgi?id=247886

Reviewed by Yusuke Suzuki.

Testing if the index is < 0 is not appropriate on 32-bit ARM.
Define and use minTypeValue instead.

Implementing https://bugs.webkit.org/show_bug.cgi?id=247454 would be a
better medium-term fix.

* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::Type::dump const):
* Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:

Canonical link: https://commits.webkit.org/257116@main
